tests/tls_handler_test.c

```diff
@@ -843,3 +843,320 @@ static int s_tls_client_channel_negotiation_success_fn(struct aws_allocator *all
+
+    /* shut down */
+    aws_channel_shutdown(incoming_args.channel, AWS_OP_SUCCESS);
+    ASSERT_SUCCESS(aws_condition_variable_wait_pred(
+        &condition_variable, &mutex, s_tls_channel_shutdown_predicate, &incoming_args));
+
+    /*no shutdown on the client necessary here (it should have been triggered by shutting down the other side). just
+     * wait for the event to fire. */
+    ASSERT_SUCCESS(aws_condition_variable_wait_pred(
+        &condition_variable, &mutex, s_tls_channel_shutdown_predicate, &outgoing_args));
+
+    /* clean up */
+    aws_client_bootstrap_release(client_bootstrap);
+    ASSERT_SUCCESS(aws_server_bootstrap_destroy_socket_listener(server_bootstrap, listener));
+    aws_server_bootstrap_release(server_bootstrap);
+    aws_tls_ctx_options_clean_up(&client_ctx_options);
+    aws_tls_connection_options_clean_up(&tls_client_conn_options);
+    aws_tls_ctx_options_clean_up(&server_ctx_options);
+    aws_tls_connection_options_clean_up(&tls_server_conn_options);
+    aws_tls_ctx_destroy(client_ctx);
+    aws_tls_ctx_destroy(server_ctx);
+    aws_host_resolver_clean_up(&resolver);
+    aws_event_loop_group_clean_up(&el_group);
+    aws_tls_clean_up_static_state();
+    return AWS_OP_SUCCESS;
+}
+AWS_TEST_CASE(tls_server_multiple_connections, s_tls_server_multiple_connections_fn)
+
+struct shutdown_listener_tester {
+    struct aws_socket *listener;
+    struct aws_server_bootstrap *server_bootstrap;
+    struct tls_test_args *outgoing_args;
+    struct aws_socket *socket;
+};
+
+static void s_shutdown_listener_task(struct aws_task *task, void *arg, enum aws_task_status status) {
+    (void)status;
+    struct shutdown_listener_tester *tester = arg;
+    /* destroy the listener */
+    aws_server_bootstrap_destroy_socket_listener(tester->server_bootstrap, tester->listener);
+    aws_mem_release(tester->outgoing_args->allocator, task);
+    AWS_FATAL_ASSERT(aws_socket_close(tester->socket) == AWS_OP_SUCCESS);
+}
+
+static void s_tester_client_connection_established_fool(struct aws_socket *socket, int error_code, void *user_data) {
+    /* connection is fooled~*/
+    (void)error_code;
+    struct shutdown_listener_tester *tester = user_data;
+    tester->socket = socket;
+
+    uint64_t run_at_ns;
+    aws_event_loop_current_clock_time(socket->event_loop, &run_at_ns);
+    run_at_ns += aws_timestamp_convert(1, AWS_TIMESTAMP_SECS, AWS_TIMESTAMP_NANOS, NULL);
+    struct aws_task *shutdown_listener_task =
+        aws_mem_acquire(tester->outgoing_args->allocator, sizeof(struct aws_task));
+    aws_task_init(shutdown_listener_task, s_shutdown_listener_task, tester, "wait_a_bit");
+    /* wait 1 sec for server side setup the channel then shut down the listener and close the socket */
+    aws_event_loop_schedule_task_future(socket->event_loop, shutdown_listener_task, run_at_ns);
+}
+
+static int s_tls_server_destroy_by_user_when_connection_is_in_processing_fn(
```

Comment:
Did you actually confirm that these tests fail when the fixes are removed?

Reply:
Yes! I pull the original code, and it will fail in the new tests